### PR TITLE
More explicit gradient state interface

### PR DESCRIPTION
**Pull request by asilvas**
Clear naming is subjective, but tried to be more explicit.

### CODE REVIEW

**Comment by yushiyangk** [comment]
Can we also rename the variable from `grad` to something else, perhaps `context` or `gradContext`. In other places `grad` refers to a Tensor instead.

**Comment by bwasti** [comment]
Thanks for bringing this up!

I agree that `context` or `ctx` would be better than `grad`.

Here's my proposal for names:

```javascript
export interface GradContext {
   forward_inputs: [Tensor, ...ArgType[]]
   forward_output: Tensor
   backward_input: Tensor // the associated gradient of forward_output
   backward_output_index: number // index of the associated forward input to be differentiated
 }
```

File: shumai/tensor/tensor_ops.ts [hunk 1]
```diff
@@ -16,7 +16,7 @@ export function relu(tensor: Tensor): Tensor {
   return tensor.maximum(scalar(0))
 }
 
-export function leakyRelu(tensor: Tensor, negative_slope: number): Tensor {
+export function leakyRelu(tensor: Tensor, negative_slope = 1e-3): Tensor {
```
Comment (bwasti) [inline shumai/tensor/tensor_ops.ts:19]: fine to sneak this in, but can you annotate with the type? `negative_slope: number = 1e-3`

**Review by bwasti — APPROVED** [review]
LGTM!  one nit with the change to leakyRelu

File: shumai/tensor/tensor_ops.ts (hunk 1 above)
Comment (asilvas) [reply shumai/tensor/tensor_ops.ts:19]: `bun format` removes the type annotation when the arg has a default

Comment (bwasti) [reply shumai/tensor/tensor_ops.ts:19]: :o @cryptodeal is that expected?

**Comment by bwasti** [comment]
Here's an updated diagram with the new names:

out

**Pull request merged**